### PR TITLE
Make CommTensor Generic to arguments and outputs structures

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84606
* #84604
* #84592
* #84585
* __->__ #84576

### CODE REVIEW

File: test/distributed/test_c10d_common.py [hunk 1]
```diff
@@ -1378,11 +1385,20 @@ class CommResult:
     _work: torch.classes.c10d.Work
 
 
-def wrap_comm_result(result):
+def wrap_comm_result(result: Tuple[Any]) -> Tuple[Any]:
+    def wrap(work, e):
+        assert isinstance(e, torch.Tensor), (
+            "Excepting collection of tensors as the first element in the "
```
Comment (wanchaol) [inline test/distributed/test_c10d_common.py:1391]: “expecting"?

File: test/distributed/test_c10d_common.py [hunk 2]
```diff
@@ -1378,11 +1385,20 @@ class CommResult:
...
+            "return value of communication operations."
+        )
+
+        return CommResult(e, work)
+
+    # E.g.,
     # allreduce_ returns ([tensor], work)
-    tensor = result[0][0]
+    # allgather_ returns ([[tensor1, tensor2]], work)
```
Comment (wanchaol) [inline test/distributed/test_c10d_common.py:1399]: any reason we return an additional [] wrapper?

**Review by aazzolini — APPROVED** [review]

File: test/distributed/test_c10d_common.py (hunk 2 above)
Comment (mrshenli) [reply test/distributed/test_c10d_common.py:1399]: it's the existing signature of allgather. When allgather was developed back in 2018, it supports single-process multi-device use cases. The outer list is for the multi-device scenarios, where which inner-list represent one device. Same thing applies to allreduce as well.

File: test/distributed/test_c10d_common.py (hunk 1 above)
Comment (mrshenli) [reply test/distributed/test_c10d_common.py:1391]: Let me update on top of the stack to avoid triggering CI again.

**Pull request closed without merging**